### PR TITLE
[SPARK-38360][SQL][SS][PYTHON] Introduce a `exists` function for `TreeNode` to eliminate duplicate code patterns

### PR DESCRIPTION
**Pull request by LuciferYang**
### What changes were proposed in this pull request?
There are many duplicate code patterns in Spark code:

```scala
treeNode.find(condition).isDefined 
treeNode.find(condition).nonEmpty 
treeNode.find(condition).isEmpty
```

This pr Introduce a `exists` function for `TreeNode` to simplify them, after this pr:

- `treeNode.find(condition).isDefined ` and `treeNode.find(condition).nonEmpty ` -> `treeNode.exists(condition)`
- `treeNode.find(condition).isEmpty` -> `!treeNode.exists(condition)`

### Why are the changes needed?
Code simplification

### Does this PR introduce _any_ user-facing change?
No.

### How was this patch tested?
Pass GA and add a new UT for new function.

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/trees/TreeNode.scala [hunk 1]
```diff
@@ -246,6 +246,16 @@ abstract class TreeNode[BaseType <: TreeNode[BaseType]] extends Product with Tre
     children.foldLeft(Option.empty[BaseType]) { (l, r) => l.orElse(r.find(f)) }
   }
 
+  /**
+   * Test whether there is [[TreeNode]] satisfies the conditions specified in `f`.
+   * The condition is recursively applied to this node and all of its children (pre-order).
+   */
+  def exists(f: BaseType => Boolean): Boolean = if (f(this)) {
```
Comment (zhengruifeng) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/trees/TreeNode.scala:253]: is it possible to make `f` a `PartialFunction[BaseType, Boolean]` and futher simplify some places like 
```
    e.exists {
      case s: ScalarSubquery => s.isCorrelated
      case _ => false
    }
```

to

```
e.exists {  case s: ScalarSubquery => s.isCorrelated    }
```

?

Comment (LuciferYang) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/trees/TreeNode.scala:253]: Like
```scala
def exists(pf: PartialFunction[BaseType, Boolean]): Boolean = {
    if (pf.applyOrElse(this, { _: Any => false})) {
      true
    } else {
      children.exists(c => pf.applyOrElse(c, { _: Any => false}))
    }
  }
```

This implementation is more like `PartialFunction#cond` as follows:

```scala
 /** Creates a Boolean test based on a value and a partial function.
   *  It behaves like a 'match' statement with an implied 'case _ => false'
   *  following the supplied cases.
   *
   *  @param  x   the value to test
   *  @param  pf  the partial function
   *  @return true, iff `x` is in the domain of `pf` and `pf(x) == true`.
   */
  def cond[T](x: T)(pf: PartialFunction[T, Boolean]): Boolean = pf.applyOrElse(x, constFalse)
```

This requires the assumption that `case _ => false`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 2]
```diff
@@ -509,7 +509,7 @@ class Analyzer(override val catalogManager: CatalogManager)
     }
 
     private def hasUnresolvedAlias(exprs: Seq[NamedExpression]) =
-      exprs.exists(_.find(_.isInstanceOf[UnresolvedAlias]).isDefined)
+      exprs.exists(_.exists(_.isInstanceOf[UnresolvedAlias]))
```
Comment (LuciferYang) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:512]: @zhengruifeng In addition, if `f` becomes `PartialFunction`, there may need to be rewritten as
```
exprs.exists(_.exists { case _: UnresolvedAlias => true})
```

Comment (zhengruifeng) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:512]: yes, If we apply this PartialFunction, call sites like this need to be rewritten.

**Comment by zhengruifeng** [comment]
cc @cloud-fan @viirya @HyukjinKwon

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/trees/TreeNode.scala (hunk 1 above)
Comment (zhengruifeng) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/trees/TreeNode.scala:253]: maybe a little simpler like this:
```
  def exists(pf: PartialFunction[BaseType, Boolean]): Boolean = {
    val lifted = pf.lift
    lifted(this).getOrElse {
      children.foldLeft(false) { (l, r) => l || r.exists(pf) }
    }
  }
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala (hunk 2 above)
Comment (LuciferYang) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:512]: OK

Comment (LuciferYang) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:512]: Let's wait for more suggestions from others

**Review by viirya — COMMENTED** [review]
Hmm, I'm not sure about how much it reduces duplicate code. Seems not too much? Looks like you only eliminate `isDefined`, `isEmpty`, `nonEmpty`.

I'm okay with the change.

**Comment by HyukjinKwon** [comment]
Yeah. I personally find `find(...).isEmpty` is easier to read vs. `exists`. But don't feel strongly too.

**Comment by LuciferYang** [comment]
> Hmm, I'm not sure about how much it reduces duplicate code. Seems not too much? Looks like you only eliminate `isDefined`, `isEmpty`, `nonEmpty`.
> 
> I'm okay with the change.

Yes, just a small optimization

**Comment by LuciferYang** [comment]
> Yeah. I personally find `find(...).isEmpty` is easier to read vs. `exists`. But don't feel strongly too.

Yes, everyone may feel different. I respect your decision

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala [hunk 3]
```diff
@@ -1690,11 +1690,10 @@ object PushPredicateThroughNonJoin extends Rule[LogicalPlan] with PredicateHelpe
    */
   private def canPushThroughCondition(plan: LogicalPlan, condition: Expression): Boolean = {
     val attributes = plan.outputSet
-    val matched = condition.find {
-      case s: SubqueryExpression => s.plan.outputSet.intersect(attributes).nonEmpty
-      case _ => false
+    condition.exists {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:1693]: this one looks a bit confusing. Does it really match the previous logic?

**Comment by cloud-fan** [comment]
looks fine as it matches the scala collection API

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala (hunk 3 above)
Comment (LuciferYang) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:1693]: This is how I derived this, the original logic same as:

```scala
condition.find {
  case s: SubqueryExpression => s.plan.outputSet.intersect(attributes).nonEmpty
  case _ => false
}.isEmpty
```
It is equivalent to

```scala
!condition.find {
  case s: SubqueryExpression => s.plan.outputSet.intersect(attributes).nonEmpty
  case _ => false
}.noEmpty
```

-->

```scala
!condition.exists {
  case s: SubqueryExpression => s.plan.outputSet.intersect(attributes).nonEmpty
  case _ => false
}
```

->
```
condition.exists {
  case s: SubqueryExpression => !s.plan.outputSet.intersect(attributes).nonEmpty
  case _ => !false
}
```

->
```
condition.exists {
  case s: SubqueryExpression => s.plan.outputSet.intersect(attributes).isEmpty
  case _ => true
}
```

Comment (LuciferYang) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:1693]: hmm... there is something wrong with this derivation process

Comment (LuciferYang) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: I revert this to:

```scala
!condition.exists {
      case s: SubqueryExpression => s.plan.outputSet.intersect(attributes).nonEmpty
      case _ => false
    }
```
@cloud-fan

**Review by cloud-fan — APPROVED** [review]

**Review by zhengruifeng — APPROVED** [review]
There are also some patterns like `treeNode.collectFirst(condition).isDefined`, but I think they can be put into a follow-up.

**Comment by LuciferYang** [comment]
> There are also some patterns like treeNode.collectFirst(condition).isDefined, but I think they can be put into a follow-up.

Thanks ~ I'll find them and add a followup

**Comment by LuciferYang** [comment]
https://github.com/LuciferYang/spark/runs/5493065701

image

Seems GA passed but hang...

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Comment by LuciferYang** [comment]
thanks all

**Pull request closed without merging**